src/charm.py

```diff
@@ -144,32 +142,30 @@ def _on_certificates_relation_broken(self, event: EventBase) -> None:
         self._delete_certificate()
         self.unit.status = BlockedStatus("Waiting for certificates relation")
 
-    def _on_certificates_relation_joined(self, event: EventBase) -> None:
-        """Generates CSR and requests new certificate."""
-        if not self._container.can_connect():
-            event.defer()
-            return
-        if not self._private_key_is_stored():
-            event.defer()
-            return
-        if self._certificate_is_stored():
-            return
+    def _get_current_provider_certificate(self) -> str | None:
+        """Compares the current certificate request to what is in the interface.
 
-        self._request_new_certificate()
+        Returns The current valid provider certificate if present
+        """
+        csr = self._get_stored_csr()
+        for provider_certificate in self._certificates.get_assigned_certificates():
+            if provider_certificate.csr == csr:
+                return provider_certificate.certificate
+        return None
 
-    def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
-        """Pushes certificate to workload and configures workload."""
-        if not self._container.can_connect():
-            event.defer()
-            return
-        if not self._csr_is_stored():
-            logger.warning("Certificate is available but no CSR is stored")
-            return
-        if event.certificate_signing_request != self._get_stored_csr():
-            logger.debug("Stored CSR doesn't match one in certificate available event")
-            return
-        self._store_certificate(event.certificate)
-        self._configure_nssf(event)
+    def _update_certificate(self, provider_certificate) -> bool:
+        """Compares the provided certificate to what is stored.
+
+        Returns True if the certificate was updated
+        """
+        existing_certificate = (
+            self._get_stored_certificate() if self._certificate_is_stored() else ""
+        )
+
+        if existing_certificate != provider_certificate:
+            self._store_certificate(certificate=provider_certificate)
+            return True
+        return False
```

Comment:
This method does 2 things, it checks of the existinc cert is the same as the provided one and it updates it. I recommend only doing 1 thing here, updating the cert (as the method name implies) and not returning a bool, especially since we don't use the return value anywhere.

Reply:
actually the return value of `_update_certificate` is used.
We also have the same pattern for `_update_config_file`

Reply:
I think we renaming the method could be helpful here. If we give a name such as `def _is_certificate_update_required` name matches with the actual duty.

Reply:
I agree to Guillaume here. The method detects if update required, then updates and returns if it's changed.  Giving single responsibility to the methods is preferred.